### PR TITLE
Transaction solidifier 

### LINKED ISSUE #214: Implement milestones tracker
Still needs to:
- [x] checkSolidity
- [x] updateSnapshot

### LINKED ISSUE #378: Implement transaction_solidifier

### LINKED ISSUE #397: Implement DAG/Tangle traversal utils
We do lots of traversals (DFS)

Provide function/s that will accept predicate (for whether to branch or not) and todo functor together with `tangle_t*`, that way, we won't have to reimplement DFS each time

### PR DESCRIPTION
**Pull request by tsvisabo**
* update schema (iota_transaction) with solid field

* implement solidifier logic

Fixes #405 
Fixes #214 
Fixes #378 
Fixes #397

### CODE REVIEW

File: common/model/transaction.h [hunk 1]
```diff
@@ -91,6 +91,7 @@ struct _iota_transaction {
 
   // Metadata
   uint64_t snapshot_index;
+  bool solid;
```
Comment (th0br0) [inline common/model/transaction.h:94]: We really need to move this metadata outside of the model. model should just be a 1:1 mapping of the trytes themselves. (same goes for the hash, that is also metadata)

Comment (th0br0) [reply common/model/transaction.h:94]: e.g. `cclient` also uses this struct, but doesn't care about the memory overhead of snapshot_index & solid

File: common/storage/sql/sqlite3/tests/test_sqlite3.c [hunk 2]
```diff
@@ -278,6 +280,7 @@ void test_update_snapshot_index(void) {
   TEST_ASSERT_EQUAL_INT(tx.snapshot_index, 123456);
 }
 
+<<<<<<< HEAD
```
Comment (th0br0) [inline common/storage/sql/sqlite3/tests/test_sqlite3.c]: remove?

File: common/storage/sql/sqlite3/tests/test_sqlite3.c [hunk 3]
```diff
@@ -333,8 +357,13 @@ int main(void) {
   RUN_TEST(test_stored_milestone);
   RUN_TEST(test_stored_load_hashes_by_address);
   RUN_TEST(test_stored_load_hashes_of_approvers);
+<<<<<<< HEAD
   RUN_TEST(test_update_snapshot_index);
   RUN_TEST(test_milestone_state_diff);
+=======
```
Comment (th0br0) [inline common/storage/sql/sqlite3/tests/test_sqlite3.c]: remove bad merge stuff

File: common/storage/sql/sqlite3/storage.c [hunk 4]
```diff
@@ -376,6 +379,31 @@ retcode_t iota_stor_transaction_load_hashes_of_approvers(
...
+  int rc = sqlite3_prepare_v2((sqlite3*)conn->db,
+                              iota_statement_transaction_update_solid_state, -1,
+                              &sqlite_statement, &err_msg);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed preparing statement, sqlite3 code is: %\" PRIu64 \"\n",
+              rc);
+    return RC_SQLITE3_FAILED_PREPARED_STATEMENT;
+  }
+  rc = column_compress_bind(sqlite_statement, 1, hash, FLEX_TRIT_SIZE_243);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: you forgot to bind `is_solid`

File: consensus/utils/tangle_traversals.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * @param tangle The tangle on which we traverse
+ * @param func The operation to do when new transaction is visited
+ * @param entry_point Where should the traversal begin from
+ * @param data Additional data, (might want to pass to func)
+ *
+ * @return error value.
+ */
+
+retcode_t tangle_traversal_dfs_to_genesis(tangle_t* const tangle,
```
Comment (th0br0) [inline consensus/utils/tangle_traversals.h:51]: is it possible to split these methods into a separate PR?

File: consensus/utils/tangle_traversals.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                      &should_stop)) != RC_OK) {
+        break;
+      }
+
+      if (should_stop) {
+        break;
+      }
+
+      if (should_branch) {
+        // TODO messageQ publish
```
Comment (th0br0) [inline consensus/utils/tangle_traversals.c]: remove this comment?

File: consensus/ledger_validator/ledger_validator.c [hunk 7]
```diff
@@ -20,52 +21,35 @@
...
-static retcode_t update_snapshot_milestone(ledger_validator_t *const lv,
-                                           flex_trit_t *const milestone_hash,
-                                           uint64_t index) {
-  retcode_t ret = RC_OK;
-  hash_stack_t non_analyzed_hashes = NULL;
-  hash_set_t analyzed_hashes = NULL;
-  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
-
-  struct _trit_array tx_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
+static retcode_t update_snapshot_milestone_do_func(
```
Comment (th0br0) [inline consensus/ledger_validator/ledger_validator.c:24]: we should also document the purpose of these static methods :thinking:

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+#define TRANSACTION_SOLIDIFIER_LOGGER_ID "consensus_transaction_solidifier"
+
+#define SOLID_PROPAGATION_INTERVAL 0.5
+
+/*
+ * Forward declarations
+ */
+
+static flex_trit_t genesis_hash[FLEX_TRIT_SIZE_243];
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c:22]: init to `FLEX_TRIT_NULL_VALUE` ?

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    if (propogate_solid_transactions(ts) != RC_OK) {
+      log_critical(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+                   "Failed in solid transaction propagation %s\n");
+    }
+    sleep(SOLID_PROPAGATION_INTERVAL);
+  }
+  return NULL;
+}
+
+static retcode_t propogate_solid_transactions(
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c]: it's `propagate`

File: common/storage/defs.h [hunk 10]
```diff
@@ -31,6 +31,7 @@
 #define TRANSACTION_COL_NONCE "nonce"
 #define TRANSACTION_COL_HASH "hash"
 #define TRANSACTION_COL_SNAPSHOT_INDEX "snapshot_index"
+#define TRANSACTION_COL_SOLID "solid"
 
 #define TRANSACTION_NUM_COLS 17
```
Comment (thibault-martinez) [inline common/storage/defs.h]: Increment ?

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+} check_solidity_do_func_params_t;
+
+static void *spawn_solid_transactions_propogation(void *arg) {
+  transaction_solidifier_t *ts = (transaction_solidifier_t *)arg;
+  while (ts->running) {
+    if (propogate_solid_transactions(ts) != RC_OK) {
+      log_critical(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+                   "Failed in solid transaction propagation %s\n");
+    }
+    sleep(SOLID_PROPAGATION_INTERVAL);
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c]: You want to use `usleep` here. Also, `SOLID_PROPAGATION_INTERVAL` is `0.5` atm

File: common/storage/sql/schema.sql [hunk 12]
```diff
@@ -15,7 +15,8 @@ CREATE TABLE IF NOT EXISTS iota_transaction (
   attachment_timestamp_upper TIMESTAMP NOT NULL,
   nonce BLOB NOT NULL,
   hash BLOB NOT NULL PRIMARY KEY,
-  snapshot_index INTEGER NOT NULL
+  snapshot_index INTEGER NOT NULL,
+  solid SMALLINT
```
Comment (thibault-martinez) [inline common/storage/sql/schema.sql]: Not null as well ?

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+void iota_consensus_transaction_solidifier_init(
+    transaction_solidifier_t *const ts, tangle_t *const tangle,
+    requester_state_t *const requester) {
+  ts->tangle = tangle;
+  ts->requester = requester;
+  ts->running = false;
+  ts->newly_set_solid_transactions = NULL;
+  ts->solid_transactions_candidates = NULL;
+  lock_handle_init(&ts->lock);
+  memset(genesis_hash, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c:129]: in general, could we move the `gensis_hash` to the tx_solidifier struct?

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+  return RC_OK;
+}
+
+void iota_consensus_transaction_solidifier_init(
+    transaction_solidifier_t *const ts, tangle_t *const tangle,
+    requester_state_t *const requester) {
+  ts->tangle = tangle;
+  ts->requester = requester;
+  ts->running = false;
+  ts->newly_set_solid_transactions = NULL;
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c:126]: you never set an actual value here? or do the HASH macros take care of this?

File: common/storage/sql/sqlite3/storage.c [hunk 15]
```diff
@@ -160,8 +160,7 @@ static retcode_t execute_statement_load_hashes(
 
 static retcode_t execute_statement_store_update(
     sqlite3_stmt* const sqlite_statement) {
-  int rc = sqlite3_step(sqlite_statement);
-
+  retcode_t rc = sqlite3_step(sqlite_statement);
```
Comment (thibault-martinez) [inline common/storage/sql/sqlite3/storage.c]: Why ? It doesn't return a `retcode_t`

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+      *solid = false;
+      return ret;
+    }
+
+    if (*solid) {
+      if ((ret = iota_tangle_transaction_update_solid_state(
+               ts->tangle, transaction->hash, true)) != RC_OK) {
+        return ret;
+      }
+      // TODO - update height
```
Comment (th0br0) [inline consensus/transaction_solidifier/transaction_solidifier.c]: Let's ignore height unless we need it elsewhere for now.

File: common/storage/sql/sqlite3/storage.c [hunk 17]
```diff
@@ -376,6 +379,33 @@ retcode_t iota_stor_transaction_load_hashes_of_approvers(
...
 }
 
+retcode_t iota_stor_transaction_update_solid_state(
+    const connection_t* const conn, flex_trit_t* const hash, bool is_solid) {
+  retcode_t ret = RC_OK;
+  sqlite3_stmt* sqlite_statement = NULL;
+
+  if ((ret = prepare_statement(
+           (sqlite3*)conn->db, &sqlite_statement,
+           iota_statement_transaction_update_snapshot_index)) != RC_OK) {
```
Comment (thibault-martinez) [inline common/storage/sql/sqlite3/storage.c]: Wrong statement

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: consensus/consensus.c [hunk 18]
```diff
@@ -70,11 +70,20 @@ retcode_t iota_consensus_init(iota_consensus_t *const consensus,
     return ret;
   }
 
+  log_info(CONSENSUS_LOGGER_ID, "Initializing transaction solidifier\n");
+  if ((ret = iota_consensus_transaction_validator_init(
```
Comment (thibault-martinez) [inline consensus/consensus.c]: Solidifier instead of validator ?

File: consensus/milestone_tracker/milestone_tracker.c [hunk 19]
```diff
@@ -217,11 +218,16 @@ static retcode_t update_latest_solid_subtangle_milestone(
     while (pack.num_loaded != 0 && milestone.index <= latest_milestone.index &&
            mt->running) {
       has_snapshot = false;
-      if (milestone.index >= mt->latest_solid_subtangle_milestone_index
-          // TODO &&
-          // transactionValidator.checkSolidity(milestoneViewModel.getHash(),
-          // true))
-      ) {
+      bool is_solid;
```
Comment (thibault-martinez) [inline consensus/milestone_tracker/milestone_tracker.c]: declare outside of the loop

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+/*
+ * Auxiliary structs
+ */
+
+typedef struct check_solidity_do_func_params_s {
+  transaction_solidifier_t *ts;
+  bool is_milestone;
+} check_solidity_do_func_params_t;
+
+static void *spawn_solid_transactions_propogation(void *arg) {
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.c]: Typo: propogation

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+  transaction_solidifier_t *ts = (transaction_solidifier_t *)arg;
+  while (ts->running) {
+    if (propogate_solid_transactions(ts) != RC_OK) {
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.c]: Typo

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 22]
```diff
@@ -0,0 +1,373 @@
...
+  }
+
+  ts->running = true;
+  log_info(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+           "Spawning transaction_solidifier thread\n");
+  if (thread_handle_create(
+          &ts->thread, (thread_routine_t)spawn_solid_transactions_propogation,
+          ts) != 0) {
+    log_critical(TRANSACTION_SOLIDIFIER_LOGGER_ID,
+                 "Spawning cIRI API thread failed\n");
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.c]: Edit

File: consensus/transaction_solidifier/transaction_solidifier.h [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+typedef struct transaction_solidifier_s {
+  tangle_t *tangle;
+  requester_state_t *requester;
+  thread_handle_t thread;
+  bool running;
+  bool solid;
+  lock_handle_t lock;
+  hash_set_t newly_set_solid_transactions;
+  hash_set_t solid_transactions_candidates;
+
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.h]: empty line

File: consensus/utils/tangle_traversals.c [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+                                          void *data) {
+  retcode_t ret = RC_OK;
+  hash_stack_t non_analyzed_hashes = NULL;
+  hash_set_t analyzed_hashes = NULL;
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
+
+  struct _trit_array tx_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
+
+  if ((ret = hash_stack_push(&non_analyzed_hashes, entry_point)) != RC_OK) {
+  }
```
Comment (thibault-martinez) [inline consensus/utils/tangle_traversals.c:28]: Do something ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: consensus/consensus.c (hunk 18 above)
Comment (tsvisabo) [reply consensus/consensus.c]: Good point

File: common/storage/sql/sqlite3/storage.c (hunk 15 above)
Comment (tsvisabo) [reply common/storage/sql/sqlite3/storage.c]: oops

File: consensus/transaction_solidifier/transaction_solidifier.c (hunk 14 above)
Comment (tsvisabo) [reply consensus/transaction_solidifier/transaction_solidifier.c:126]: They do

File: consensus/transaction_solidifier/transaction_solidifier.c (hunk 13 above)
Comment (tsvisabo) [reply consensus/transaction_solidifier/transaction_solidifier.c:129]: yes, but that means that everyone who uses it depends on that initialization

File: consensus/utils/tangle_traversals.h (hunk 5 above)
Comment (tsvisabo) [reply consensus/utils/tangle_traversals.h:51]: I agree it should have been made in a separate one, but That will be  a lot of mess to do that now

File: consensus/transaction_solidifier/transaction_solidifier.c (hunk 8 above)
Comment (tsvisabo) [reply consensus/transaction_solidifier/transaction_solidifier.c:22]: that has to be memset-ed

File: common/model/transaction.h (hunk 1 above)
Comment (tsvisabo) [reply common/model/transaction.h:94]: how do you think we should do that?
Another struct that contains transaction and metadata which will be the one to actually go into db?

Comment (th0br0) [reply common/model/transaction.h:94]: probably something like that, yeah... `struct {iota_transaction tx; uint64_t snapshot_index; ...} `

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Reset `ciri.db` please

File: common/model/transaction.c [hunk 25]
```diff
@@ -123,7 +123,6 @@ size_t transaction_deserialize_trits(iota_transaction_t transaction,
   curl_digest(tx_trits, NUM_TRITS_SERIALIZED_TRANSACTION, hash, &curl);
   flex_trits_from_trits(transaction->hash, NUM_TRITS_HASH, hash, NUM_TRITS_HASH,
                         NUM_TRITS_HASH);
-  transaction->snapshot_index = 0;
```
Comment (thibault-martinez) [inline common/model/transaction.c:126]: What if we are not calling `transaction_new` but only deserialising from trits to a statically allocated object ?
Anyway this will disappear when we'll have a metadata structure.

File: common/storage/storage.h [hunk 26]
```diff
@@ -16,6 +16,7 @@
 #include "common/storage/pack.h"
 #include "common/trinary/trit_array.h"
 #include "consensus/snapshot/state_delta.h"
+#include "utils/hash_containers.h"
```
Comment (thibault-martinez) [inline common/storage/storage.h]: Why for ?

File: consensus/consensus.c [hunk 27]
```diff
@@ -78,14 +78,21 @@ retcode_t iota_consensus_init(iota_consensus_t *const consensus,
            (testnet ? SNAPSHOT_CONF_TESTNET : SNAPSHOT_CONF_MAINNET),
            testnet)) != RC_OK) {
     log_critical(CONSENSUS_LOGGER_ID, "Initializing snapshot failed\n");
+  }
```
Comment (thibault-martinez) [inline consensus/consensus.c:81]: new line

File: consensus/milestone_tracker/milestone_tracker.c [hunk 28]
```diff
@@ -218,11 +220,15 @@ static retcode_t update_latest_solid_subtangle_milestone(
...
       has_snapshot = false;
-      if (milestone.index >= mt->latest_solid_subtangle_milestone_index
-          // TODO &&
-          // transactionValidator.checkSolidity(milestoneViewModel.getHash(),
-          // true))
-      ) {
+      if (milestone.index >= mt->latest_solid_subtangle_milestone_index) {
+        if ((ret = iota_consensus_transaction_solidifier_check_solidity(
+                 mt->transaction_solidifier, milestone.hash, true,
+                 &is_solid))) {
```
Comment (thibault-martinez) [inline consensus/milestone_tracker/milestone_tracker.c]:
```suggestion
                 &is_solid)) != RC_OK) {
```

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 29]
```diff
@@ -0,0 +1,383 @@
...
+    }
+    usleep(SOLID_PROPAGATION_INTERVAL);
+  }
+  return NULL;
+}
+
+static retcode_t propagate_solid_transactions(
+    transaction_solidifier_t *const ts) {
+  retcode_t ret;
+  hash243_set_t transactions_to_propagate;
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.c]:
```suggestion
  hash243_set_t transactions_to_propagate = NULL;
```

File: consensus/transaction_solidifier/transaction_solidifier.c [hunk 30]
```diff
@@ -0,0 +1,383 @@
...
+  lock_handle_lock(&ts->lock);
+  if ((ret = hash243_set_for_each(&ts->newly_set_solid_transactions,
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.c:72]: You can also use `hash243_set_append` on a `NULL` set

Comment (thibault-martinez) [reply consensus/transaction_solidifier/transaction_solidifier.c:72]: That make me think that we can re-implement `hash243_set_append` with `hash243_set_for_each `

File: consensus/transaction_solidifier/transaction_solidifier.h [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+extern retcode_t iota_consensus_transaction_solidifier_check_solidity(
+    transaction_solidifier_t *const ts, flex_trit_t *const hash,
+    bool is_milestone, bool *const is_solid);
+
+extern retcode_t
+iota_consensus_transaction_solidifier_check_and_update_solid_state(
+    transaction_solidifier_t *const ts, flex_trit_t *const hash);
+
+// TODO - complete
```
Comment (thibault-martinez) [inline consensus/transaction_solidifier/transaction_solidifier.h]: What's missing ?

File: consensus/utils/tangle_traversals.c [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+retcode_t tangle_traversal_dfs_to_genesis(tangle_t *const tangle,
+                                          tangle_traversal_functor func,
+                                          flex_trit_t *const entry_point,
+                                          void *data) {
+  retcode_t ret = RC_OK;
+  hash243_stack_t non_analyzed_hashes = NULL;
+  hash243_set_t analyzed_hashes = NULL;
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, pack);
+
+  struct _trit_array tx_hash = {NULL, NUM_TRITS_HASH, FLEX_TRIT_SIZE_243, 0};
```
Comment (thibault-martinez) [inline consensus/utils/tangle_traversals.c]: > If you're initialising a struct, please use member designators to identify the individual fields

File: consensus/utils/tangle_traversals.c (hunk 6 above)
Comment (thibault-martinez) [reply consensus/utils/tangle_traversals.c]: Comment still here

File: gossip/components/processor.c [hunk 33]
```diff
@@ -110,6 +110,11 @@ static retcode_t process_request_bytes(processor_state_t *const state,
     // If requested hash is equal to transaction hash: request a random tip
     trit_array_set_null(request_hash);
   }
+
+  if ((ret = iota_consensus_transaction_solidifier_check_and_update_solid_state(
+           &state->node->core->consensus.transaction_solidifier, (*tx).hash))) {
```
Comment (thibault-martinez) [inline gossip/components/processor.c]:
```suggestion
           &state->node->core->consensus.transaction_solidifier, tx->hash))) {
```

File: utils/hash_containers.h [hunk 34]
```diff
@@ -19,6 +19,73 @@
...
+
+uint32_t hash_set_size(hash_set_t const *const set);
+retcode_t hash_set_add(hash_set_t *const set, flex_trit_t const *const hash);
+retcode_t hash_set_append(hash_set_t const *const set, hash_set_t *const clone);
+bool hash_set_contains(hash_set_t const *const set,
+                       flex_trit_t const *const hash);
+void hash_set_free(hash_set_t *const set);
+retcode_t hash_set_for_each(hash_set_t *set, hash_foreach_func func,
+                            void *container);
+
```
Comment (thibault-martinez) [inline utils/hash_containers.h]: Remove them ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: common/storage/storage.h (hunk 26 above)
Comment (tsvisabo) [reply common/storage/storage.h]: inadvartent

**Comment by tsvisabo** [comment]
Fixes #397 #378

**Review by thibault-martinez — APPROVED** [review]

**Review by th0br0 — APPROVED** [review]

File: consensus/transaction_solidifier/transaction_solidifier.c (hunk 8 above)
Comment (th0br0) [reply consensus/transaction_solidifier/transaction_solidifier.c:22]: it doesn't actually.
